lint/dagger/src/main/java/dev/whosnickdoglio/dagger/detectors/ScopedWithoutInjectAnnotationDetector.kt

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2023 Nicholas Doglio
+ * SPDX-License-Identifier: MIT
+ */
+package dev.whosnickdoglio.dagger.detectors
+
+import com.android.tools.lint.checks.getFqName
+import com.android.tools.lint.client.api.UElementHandler
+import com.android.tools.lint.detector.api.Category
+import com.android.tools.lint.detector.api.Detector
+import com.android.tools.lint.detector.api.Implementation
+import com.android.tools.lint.detector.api.Issue
+import com.android.tools.lint.detector.api.JavaContext
+import com.android.tools.lint.detector.api.Scope
+import com.android.tools.lint.detector.api.Severity
+import com.android.tools.lint.detector.api.SourceCodeScanner
+import com.android.tools.lint.detector.api.TextFormat
+import dev.whosnickdoglio.lint.shared.INJECT
+import dev.whosnickdoglio.lint.shared.SCOPE
+import org.jetbrains.uast.UAnnotated
+import org.jetbrains.uast.UClass
+import org.jetbrains.uast.UElement
+import org.jetbrains.uast.resolveToUElement
+
+/**
+ * A Lint rule that warns if a class is annotated with any scope annotation but does not have a
+ * `@Inject` annotation on any constructor that it will not be added to the Dagger graph.
+ */
+internal class ScopedWithoutInjectAnnotationDetector : Detector(), SourceCodeScanner {
+
+    override fun getApplicableUastTypes(): List<Class<out UElement>> = listOf(UClass::class.java)
+
+    override fun createUastHandler(context: JavaContext): UElementHandler =
+        object : UElementHandler() {
+            override fun visitClass(node: UClass) {
+                val sourceScopeAnnotations =
+                    node.uAnnotations
+                        .map { annotation -> annotation.resolveToUElement() }
+                        .filterIsInstance<UAnnotated>()
+                        .filter { annotated ->
+                            annotated.uAnnotations.any { annotation ->
+                                annotation.qualifiedName == SCOPE
+                            }
+                        }
+                        .map { it.javaPsi }
+
+                val scopeAnnotationsOnCurrentClass =
+                    node.uAnnotations.filter { annotation ->
+                        sourceScopeAnnotations.any { scope ->
+                            scope?.getFqName() == annotation.qualifiedName
+                        }
+                    }
+
+                val isInjected =
+                    node.constructors.any { constructor -> constructor.hasAnnotation(INJECT) }
+
+                if (scopeAnnotationsOnCurrentClass.isNotEmpty() && !isInjected) {
+                    scopeAnnotationsOnCurrentClass.forEach { annotation ->
+                        context.report(
+                            issue = ISSUE,
+                            location = context.getLocation(annotation),
+                            message = ISSUE.getExplanation(TextFormat.TEXT),
+                            quickfixData =
```

Comment:
I thought about adding a quickfix for adding an `@Inject` to the constructor but wasn't sure if that was a good idea if class in question took parameters that weren't already on the DI graph. 🤔

Reply:
Agree, I think that the safest bet is to remove the Scope. I mean, if the code was compiling without the `@Inject` there is no point to add it, just remove the `@Scope` to make the code easier to read/maintain.